### ISSUE TITLE
Status: TCP conn broken (Help)

### ISSUE THREAD

**Issue by breezelunar** [uhi22/pyPLC#35]
[BMW-Logs.pdf](https://github.com/user-attachments/files/17869370/BMW-Logs.pdf)
[MINI-COOPER-LOGS.pdf](https://github.com/user-attachments/files/17869371/MINI-COOPER-LOGS.pdf)

Hi uhi22

I have been testing the pyPLC code with a pi5, a PLC Stamp Micro and two vehicles, a MINI Cooper SE and a BMW ix3

I suspect the EV is suspending communication during the supportedAppProtocolReq/-Res, I was hoping you might be able to advise on what is causing the TCP connection to break.

I appreciate your work and your time, thank you in advance.

-breezelunar

**Comment by breezelunar** [comment]
Hi uhi22, thank you for the swift reply. Initially I thought it could be the CP line. But the CP line has been tested rigorously with an oscilloscope, I don't believe the physical connection is the issue. I have also disabled the network manager for our specific eth1 port. Any other ideas?

**Comment by breezelunar** [comment]
It seems to happen at the same place every time, the supportedAppProtocolReq/-Res. Could this be a clue?

**Comment by uhi22** [comment]
We would need the support of a linux expert who knows how view the reason why a TCP connection just closes. First interesting point is a wireshark log of the interface, this should show whether the car or the raspberry closes the TCP.

**Comment by breezelunar** [comment]
Thanks uhi, I will check out that wireshark log

**Comment by breezelunar** [comment]
Could this be the fix? Let me know

https://github.com/uhi22/pyPLC/issues/9

**Comment by breezelunar** [comment]
Does this possibly relate to issue #9 at all?

https://github.com/uhi22/pyPLC/issues/9

**Comment by breezelunar** [comment]
Does this possibly relate to issue https://github.com/uhi22/pyPLC/issues/9 at all?

**Comment by uhi22** [comment]
At 6.14s there is a FIN, this is the point where the TCP connection is closing.

**Comment by uhi22** [comment]
Could you attach the real pcap file?

**Comment by breezelunar** [comment]
[capture.pcap.zip](https://github.com/user-attachments/files/17919147/capture.pcap.zip)

**Comment by breezelunar** [comment]
Thank you very much uhi for the analysis. Based on the information you provided, I am going to do some work to solve the issue, starting with the PWM on the CP line. Will let you know my findings. Thanks again for your time

**Comment by lewurm** [comment]
I had a similar issue with my car (Tesla Model 3).  My problem was indeed the CP line. There is a state change from B to C somewhen soon after the schema selection (i.e. 9V to 6V). That messed with my (faulty) circuit, the solution was to swap the LF353 with a genuine one; you could see with the scope that the slew rate wasn't proper before. The car even gave  somewhat useful errors in its service menu.

**Comment by breezelunar** [comment]
lewurm, what did you use as a replacement for the LF353?

**Comment by uhi22** [comment]
My understanding is that he used a "faked" Chip or an other type before, and changing to a real LF353 solved the issue.

**Comment by lewurm** [comment]
That's right, I used fake LF353 from Aliexpress before. Using a genuine TI LF353 fixed my issue around broken connections after schema selection (that is, the voltage levels on CP looked better then in state C).

**Comment by uhi22** [comment]
So my conclusion: If something does not work, take an oscilloscope to measure the CP. This should show nice rectangles of the 1kHz, 5% PWM, with the low voltage nearly at -12V, and the high voltage near 9V (in state B) or later near 6V (in state C).

**Comment by breezelunar** [comment]
Yes it seems the chargepump I was using to make the negative rail could be the issue. It only reaches -11.8V. So I am working on a new design now. Otherwise the PWM looks good. No noise, very clean. So will keep updated. Thanks again for the help

**Comment by breezelunar** [comment]
I get 12.1V on the positive and about -10.5V on the negative (these are the scope measurements, RIGOL DS1104), I measure -11.8V on the output of the chargepump, so there is even more loss for some reason on the negative level of the CP. 

I have a new circuit now, using MUN5312s and BSS8402s for the 12V/-12V rails.

**Comment by uhi22** [comment]
Good point. The nominal voltages are -12V and +9V in state B. And the reality on my setup is -10V and +7V. So two volts are missing in each direction. Nevertheless this worked on some cars. A better (output rail to rail) amplifier would be great, to be fully in spec.
![image](https://github.com/user-attachments/assets/e0cb750e-b38b-4654-99d4-66bc2dc63694)

**Comment by breezelunar** [comment]
That's very interesting that even with a negative rail of -10V you are still able to communicate

**Comment by uhi22** [comment]
Some people are even saying that some onboard chargers are even happy when using 0V as lower voltage, so just applying a 0V / 12V PWM :-)

**Comment by breezelunar** [comment]
![Uploading Screenshot 2025-01-10 at 10.44.24.png…]()

**Comment by breezelunar** [comment]
> Some people are even saying that some onboard chargers are even happy when using 0V as lower voltage, so just applying a 0V / 12V PWM :-)

Wow, but then I wonder if the rails are really the issue in my case. Perhaps the rise time is the issue?

**Comment by breezelunar** [comment]
Hi Uhi22, just two quick questions, what is the incremental rate of increase of the EVSE present voltage during the preCharge phase? Also, at what phase in the Req/Res communication do the actual positive and negative battery terminals make themselves available on the cable? Allowing current to flow?

Thanks

**Comment by breezelunar** [comment]
Just an update, we are now able to charge our car at 10kW using an MPPT directly from our solar panels! 

Thanks for all the help, will keep you updated with further developments

**Comment by uhi22** [comment]
No further open question here. Closing.
